[RF] Add new `RooExpPoly` and `RooPower` PDFs

cburgard opened the pull request:
At the moment, it's not really possible to implement a power law or a exponential of a polynomial (as used, for example, in the ATLAS H->yy analysis) without using RooGenericPdf.
This PR adds two new Pdf classes that implement an exponential polynomial and a power law.

cburgard commented:
@hageboeck @lmoneta : these are super simple additional classes. following up on our discussion on the "RooAbsPdf" vs. "RooAbsReal" business: do you think it would be useful for me to template these classes in such a way as to meet my suggestion of avoiding code duplication?

oshadura commented:
@phsft-bot build!

hageboeck commented on roofit/roofit/src/RooExpPoly.cxx:
```diff
@@ -0,0 +1,148 @@
+  RooAbsArg* coef ;
+  while((coef = (RooAbsArg*)coefIter.next())) {
+    if (!dynamic_cast<RooAbsReal*>(coef)) {
+      coutE(InputArguments) << "RooExpPoly::ctor(" << GetName() << ") ERROR: coefficient " << coef->GetName()
```
Let's remove the (slower) legacy iterators:
`for (const auto coef : coefList) {`

hageboeck commented on roofit/roofit/src/RooExpPoly.cxx:
```diff
@@ -0,0 +1,148 @@
+    RooFIter it = _coefList.fwdIterator();
+    RooAbsReal* c;
+    while ((c = (RooAbsReal*) it.next())) coefs.push_back(c->getVal(nset));
+  }
```
```
for (const auto coef : _coefList) {
  auto c = static_cast<RooAbsReal*>(c);
```

hageboeck commented on roofit/roofit/src/RooPower.cxx:
```diff
@@ -0,0 +1,154 @@
+  RooFIter coefIter = coefList.fwdIterator() ;
+  RooFIter expIter = expList.fwdIterator() ;
+  RooAbsArg* coef, * exp ;
+  while((coef = (RooAbsArg*)coefIter.next()) && (exp = (RooAbsArg*)expIter.next())) {
```
Here, a counting loop should be faster:
```
for (unsigned int i=0; i < coefList.size(); ++i) {
   auto coef = coefList[i];
   auto exp  = expList[i];

hageboeck commented on roofit/roofit/src/RooPower.cxx:
```diff
@@ -0,0 +1,154 @@
+    while ((c = (RooAbsReal*) coef_it.next())) coefs.push_back(c->getVal(nset));
+    RooFIter exp_it = _expList.fwdIterator();
+    while ((c = (RooAbsReal*) exp_it.next())) exps.push_back(c->getVal(nset));
+    
```
Range-based for loops should do the trick.

hageboeck commented:
Hi Carsten, looks great so far. Three things, one optional one:
- I left some comments to modernise the iterators.
- Did I miss something or is RooPower implementing the same as RooPolynomial?
- I believe we have a solution to resolve the PDF-Function duality, but that is better discussed in person.
- [Optional, but nice to have] Could you send me plots of the two? I would like to include them on the webpage and try to make it look like this:
https://root.cern.ch/doc/master/classRooJohnson.html

cburgard commented:
Thanks for all the suggestions. I implemented all of them (hopefully) and also added an analytical integral for RooPower. For RooExpPoly, I don't think analytical integrals can be implemented easily, unless for a few special cases (only order-1 terms), in which case the user is better-off using RooExponential.
RooPower does *not* do the same thing as RooPolynomial, the main difference being that it allows for non-integer and negative exponents, which can also be derived RooAbsReal Objects (functions, fit parameters) themselves, which RooPolynomial cannot do.
Let's indeed discuss the duality issue in person.

Here's some code to make the documentation plots you asked for.

For RooExpPoly:

    import ROOT
    
    ws = ROOT.RooWorkspace("ws")
    ws.factory("x[1.,-10,10]")
    
    poly0 = ws.factory("RooExpPoly::poly0(x,{1})")
    poly1 = ws.factory("RooExpPoly::poly1(x,{-1},2)")
    
    frame = ws.var("x").frame(ROOT.RooFit.Title("RooExpPoly"))
    c = ROOT.TCanvas("canvas")
    poly0.plotOn(frame,ROOT.RooFit.Name("poly0_curve"),ROOT.RooFit.LineColor(ROOT.kRed))
    poly1.plotOn(frame,ROOT.RooFit.Name("poly1_curve"),ROOT.RooFit.LineColor(ROOT.kBlue))
    frame.Draw()
    frame.GetYaxis().SetTitle("f(x)")
    
    leg = ROOT.TLegend(0.2,0.5,0.4,0.9)
    leg.SetFillStyle(0)
    leg.SetLineWidth(0)
    leg.AddEntry(frame.getCurve("poly0_curve"),"exp(x)","l")
    leg.AddEntry(frame.getCurve("poly1_curve"),"exp(-x^{2})","l")
    leg.Draw()
    
    c.SaveAs("exppoly.pdf")

For RooPower:

    import ROOT
    
    ws = ROOT.RooWorkspace("ws")
    ws.factory("x[1.,0.3,10]")
    
    pow0 = ws.factory("RooPower::pow0(x,{1.},{-1})")
    pow1 = ws.factory("RooPower::pow1(x,{1.},{0.5})")
    pow2 = ws.factory("RooPower::pow2(x,{1.},{2.})")
    pow3 = ws.factory("RooPower::pow3(x,{2.,-0.6},{2.,2.5})")
    
    frame = ws.var("x").frame(ROOT.RooFit.Title("RooPower"))
    c = ROOT.TCanvas("canvas")
    pow0.plotOn(frame,ROOT.RooFit.Name("pow0_curve"),ROOT.RooFit.LineColor(ROOT.kRed))
    pow1.plotOn(frame,ROOT.RooFit.Name("pow1_curve"),ROOT.RooFit.LineColor(ROOT.kGreen))
    pow2.plotOn(frame,ROOT.RooFit.Name("pow2_curve"),ROOT.RooFit.LineColor(ROOT.kBlue))
    pow3.plotOn(frame,ROOT.RooFit.Name("pow3_curve"),ROOT.RooFit.LineColor(ROOT.kYellow))
    frame.Draw()
    frame.GetYaxis().SetTitle("f(x)")
    
    leg = ROOT.TLegend(0.5,0.5,0.9,0.9)
    leg.SetFillStyle(0)
    leg.SetLineWidth(0)
    leg.AddEntry(frame.getCurve("pow0_curve"),"x^{-1}","l")
    leg.AddEntry(frame.getCurve("pow1_curve"),"x^{0.5}","l")
    leg.AddEntry(frame.getCurve("pow2_curve"),"x^{2}","l")
    leg.AddEntry(frame.getCurve("pow3_curve"),"2x^{2}-0.5x^{2.5}","l")
    leg.Draw()
    
    c.SaveAs("pow.pdf")

hageboeck commented:
@phsft-bot build

hageboeck commented on roofit/roofit/src/RooExpPoly.cxx:
```diff
@@ -139,5 +139,82 @@ Double_t RooExpPoly::evaluate() const
+{
+  return RooAbsPdf::getLogVal(nset);
+  return this->evaluateLog();
+}
```
I guess this wasn't the intention. Only one of the two should stay.

hageboeck commented on roofit/roofit/src/RooExpPoly.cxx:
```diff
@@ -139,5 +139,82 @@ Double_t RooExpPoly::evaluate() const
+////////////////////////////////////////////////////////////////////////////////
+
+#define PI 3.1415926
+
```
Maybe TMath::Pi()?

oshadura commented:
@phsft-bot build!

hageboeck commented:
@phsft-bot build

hageboeck commented:
@phsft-bot build

hageboeck commented:
Hi @cburgard,

Oksana pointed me to this PR. I had asked you about plots in the documentation, so I added them. I also fixed the override warning from the last CI run. I pushed the two commits directly to your branch, so check please if we are ready to merge.

hageboeck commented:
@phsft-bot build

cburgard commented:
In the (long) time since this PR was created, a lot of debugging has happened. I have pushed a new version that is relatively stable. Please re-review!

hageboeck commented:
@phsft-bot build

hageboeck commented:
@phsft-bot build

hageboeck commented on roofit/roofit/src/RooExpPoly.cxx:
```diff
@@ -0,0 +1,311 @@
+
+**/
+
+#include <cmath>
+#include <sstream>
+#include <cassert>
+
+#include "TMath.h"
+
+#include "RooExpPoly.h"
+#include "RooAbsReal.h"
+#include "RooArgList.h"
+#include "RooMsgService.h"
+
+#include "TError.h"
```
This might fix the compile errors. (Not sure if I got it all.)
```suggestion
#include "RooExpPoly.h"

#include "RooAbsReal.h"
#include "RooArgList.h"
#include "RooMsgService.h"
#include "RooRealVar.h"

#include "TMath.h"
#include "TError.h"

#include <cmath>
#include <sstream>
#include <cassert>
#include <complex>
```

hageboeck commented:
@phsft-bot build

guitargeek commented:
@phsft-bot build
